pylabrobot/resources/volume_tracker.py

```diff
@@ -97,6 +134,18 @@ def add_liquid(self, liquid: Optional["Liquid"], volume: float) -> None:
       raise TooLittleVolumeError(
         f"Container has too little volume: {volume}uL > {self.get_free_volume()}uL.")
 
+    # Update the liquid history tracker if needed
+    if not self._is_cross_contamination_tracking_disabled:
+      # If this is a new liquid, and if the tip isn't clean
+      if liquid is not None and (liquid not in self.liquid_history) and self.liquid_history:
+        raise CrossContaminationError(
+            f"Attempting to aspirate {liquid} with "
+            f"a tip contaminated with {self.liquid_history}.")
```

Comment:
this code will actually also be called on the well when dispensing liquid into it. In that case, this error should not be raised (= it should be possible to dispense a new liquid into a well without raising). Perhaps I was wrong in suggesting this code go here, and your initial idea to put it in LiquidHandler was better.